File: retail_report/retail_report/report/product_bundle_report/product_bundle_report.js
// Copyright (c) 2023, Alimerdan Rahimov and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Product Bundle Report"] = {
	"filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
        },
    ],

    "formatter": function (value, row, column, data, default_formatter) {
        if (column.fieldname == "product_1" || column.fieldname == "product_2") {
            return `<a href="#Form/Item/${data.item_code}" target="_blank">${value}</a>`;
        }
        else {
            return default_formatter(value, row, column, data);
        }
    },

    "onload": function (report) {
        report.page.add_inner_button(__('Export to CSV'), function () {
            report.export_report('CSV');
        });
    },

    "refresh": function (report) {
        var filters = report.get_values();
        frappe.call({
            method: "myapp.myreport.execute",
            args: filters,
            callback: function (data) {
                var columns = [
                    { "label": __("Product 1"), "fieldname": "product_1", "fieldtype": "Link", "options": "Item" },
                    { "label": __("Product 2"), "fieldname": "product_2", "fieldtype": "Link", "options": "Item" },
					{ "label": __("Number of Orders"), "fieldname": "orders", "fieldtype": "Int" },
					{ "label": __("Total Amount"), "fieldname": "total_amount", "fieldtype": "Currency" },
					];
					report.set_columns(columns);
					report.set_data(data.message);
					}
					});
					}
					};

File: retail_report/retail_report/report/product_bundle_report/product_bundle_report.py
from frappe import _
import frappe 
def execute(filters=None):
    data = []
    columns = [
        _("Product 1"),
        _("Product 2"),
        _("Number of Orders"),
        _("Total Amount"),
    ]

    query = """
        SELECT t1.item_name AS product_1, t2.item_name AS product_2, COUNT(*) AS orders, SUM(t1.amount + t2.amount) AS total_amount
        FROM `tabSales Invoice Item` t1
        INNER JOIN `tabSales Invoice Item` t2 ON t1.parent = t2.parent AND t1.item_code < t2.item_code
        WHERE t1.docstatus = 1 AND t2.docstatus = 1
        GROUP BY t1.item_code, t2.item_code
        HAVING orders > 1
        ORDER BY orders DESC
    """

    results = frappe.db.sql(query, as_dict=True)

    for row in results:
        data.append([
            row.product_1,
            row.product_2,
            row.orders,
            row.total_amount,
        ])

    return columns, data